src/arch/ucx/machine.C

```diff
@@ -0,0 +1,583 @@
+
+    UcxPrepostRxBuffers();
+
+    // Ensure connects completion
+    status = ucp_worker_flush(ucxCtx.worker);
+    UCX_CHECK_STATUS(status, "ucp_worker_flush");
+
+#if CMK_SMP
+    ucxCtx.txQueue = PCQueueCreate();
+#endif
+
+    UCX_LOG(5, "Initialized: preposted reqs %d, rndv thresh %d\n",
+            ucxCtx.numRxReqs, ucxCtx.eagerSize);
+}
+
+static inline UcxRequest* UcxPostRxReq(ucp_tag_t tag, size_t size,
+                                       ucp_tag_message_h msg)
+{
+    void *buf = CmiAlloc(size);
+    UcxRequest *req;
+
+    if (tag == UCX_MSG_TAG_DIRECT) {
+        req = (UcxRequest*)ucp_tag_recv_nb(ucxCtx.worker, buf,
+                                           ucxCtx.eagerSize,
+                                           ucp_dt_make_contig(1), tag,
+                                           UCX_MSG_TAG_MASK,
+                                           UcxRxReqCompleted);
+    } else {
+        CmiEnforce(tag == UCX_MSG_TAG_PROBE);
+        req = (UcxRequest*)ucp_tag_msg_recv_nb(ucxCtx.worker, buf, size,
+                                               ucp_dt_make_contig(1), msg,
+                                               UcxRxReqCompleted);
+    }
+
+    CmiEnforce(!UCS_PTR_IS_ERR(req));
+    UCX_LOG(3, "Posted RX buf %p size %zu, req %p, tag %zu, comp %d\n",
+            req->msgBuf, size, req, tag, req->completed);
+
+    // Request completed immediately
+    if (req->completed) {
+        int idx = req->idx;
+
+        if (!(tag & UCX_RMA_TAG_MASK)) {
+            handleOneRecvedMsg(size, (char*)buf);
+        }
+
+        UCX_REQUEST_FREE(req);
+
+        if (tag & UCX_MSG_TAG_DIRECT) {
+            ucxCtx.rxReqs[idx] = UcxPostRxReq(UCX_MSG_TAG_DIRECT, ucxCtx.eagerSize, NULL);
+            ucxCtx.rxReqs[idx]->idx = idx;
+            req = ucxCtx.rxReqs[idx];
+        } else {
+            return NULL;
+        }
+    } else {
+        req->msgBuf = buf;
+    }
+
+    return req;
+}
+
+static inline UcxRequest* UcxHandleRxReq(UcxRequest *request, char *rxBuf,
+                                         size_t size, ucp_tag_t tag, int idx)
+{
+    if (!(tag & UCX_RMA_TAG_MASK)) {
+        handleOneRecvedMsg(size, rxBuf);
+    }
+
+    UCX_REQUEST_FREE(request);
+
+    if (tag & UCX_MSG_TAG_DIRECT) {
+        ucxCtx.rxReqs[idx]      = UcxPostRxReq(UCX_MSG_TAG_DIRECT,
+                                               ucxCtx.eagerSize, NULL);
+        ucxCtx.rxReqs[idx]->idx = idx;
+        return ucxCtx.rxReqs[idx];
+    }
+
+    return NULL;
+}
+
+static void UcxRxReqCompleted(void *request, ucs_status_t status,
+                              ucp_tag_recv_info_t *info)
+{
+    UcxRequest *req = (UcxRequest*)request;
+
+    UCX_LOG(3, "status %d len %d, buf %p, req %p, tag %zu\n",
+            status,  info->length, req->msgBuf, request, info->sender_tag);
+
+    if (ucs_unlikely(status == UCS_ERR_CANCELED)) {
+        return;
+    }
+
+#if CMK_ONESIDED_IMPL
+    if (info->sender_tag & UCX_RMA_TAG_PUT) {
```

Comment:
Inside LrtsIssueRget, when a source buffer is unregistered, it is sent back to the source with the tag 'UCX_RMA_TAG_PUT'. However, I see that on receiving this message, you directly perform a PUT using UcxRmaOp. Isn't it required to register the source buffer using ucp_mem_map before issuing the PUT? 

(Same question for receiving UCX_RMA_TAG_GET and performing a GET).